File: Observer.java
/*****
 * Observer
 * Define a one-to-many dependency so that when one object changes state,
 * all its dependents are notified and updated automatically.
*****/

import java.util.List;
import java.util.ArrayList;


interface Observer_ { void update(int state); }

class ConcreteObserver implements Observer_ { public void update(int state) { System.out.println(state); } }

class Subject {
    private List<Observer_> Observers = new ArrayList<>();
    private int state = 0;

    public void attach(Observer_ o) { Observers.add(o); }

    public void setState(int newState) {
        state = newState;
        for (Observer_ Observer_ : Observers) Observer_.update(state);
    }
}


public class Observer {
    public static void main(String[] args) {
        Subject s = new Subject();
        ConcreteObserver o = new ConcreteObserver();

        s.attach(o);
        s.setState(10);
        s.setState(20);
    }
}

File: Observer.cpp
/*****
 * Observer
 * Define a one-to-many dependency so that when one object changes state,
 * all its dependents are notified and updated automatically.
*****/

#include <iostream>
#include <memory>
#include <vector>

using namespace std;


class Observer { public: virtual void update(int state) = 0; };

class ConcreteObserver : public Observer { public: void update(int state) override { cout << state << endl; } };

class Subject {
    vector<Observer*> observers;
    int state = 0;

    public:
        void attach(Observer* o) { observers.push_back(o); }
        void setState(int new_state) {
            state = new_state;
            for (auto observer : observers) observer->update(state);
        }
};


int main() {
    Subject s;
    ConcreteObserver o;

    s.attach(&o);
    s.setState(10);
    s.setState(20);

    return 0;
}
